File: mobile/lib/shared/models/org_subscription.dart
import 'organization.dart';
import 'plan.dart';

class OrgSubscription {
  final String id;
  final String orgId;
  final String planId;
  final String status;
  final String? stripeCustomerId;
  final String? stripeSubscriptionId;
  final DateTime? currentPeriodEnd;
  final DateTime createdAt;
  final DateTime updatedAt;
  final DateTime? deletedAt;
  final Organization org;
  final Plan plan;

  const OrgSubscription({
    required this.id,
    required this.orgId,
    required this.planId,
    required this.status,
    this.stripeCustomerId,
    this.stripeSubscriptionId,
    this.currentPeriodEnd,
    required this.createdAt,
    required this.updatedAt,
    this.deletedAt,
    required this.org,
    required this.plan,
  });

  factory OrgSubscription.fromJson(Map<String, dynamic> json) {
    return OrgSubscription(
      id: json['id'] as String,
      orgId: json['orgId'] as String,
      planId: json['planId'] as String,
      status: json['status'] as String,
      stripeCustomerId: json['stripeCustomerId'] as String?,
      stripeSubscriptionId: json['stripeSubscriptionId'] as String?,
      currentPeriodEnd: json['currentPeriodEnd'] != null ? DateTime.parse(json['currentPeriodEnd'] as String) : null,
      createdAt: DateTime.parse(json['createdAt'] as String),
      updatedAt: DateTime.parse(json['updatedAt'] as String),
      deletedAt: json['deletedAt'] != null ? DateTime.parse(json['deletedAt'] as String) : null,
      org: Organization.fromJson(json['org'] as Map<String, dynamic>),
      plan: Plan.fromJson(json['plan'] as Map<String, dynamic>),
    );
  }

  Map<String, dynamic> toJson() {
    return {
      'id': id,
      'orgId': orgId,
      'planId': planId,
      'status': status,
      'stripeCustomerId': stripeCustomerId,
      'stripeSubscriptionId': stripeSubscriptionId,
      'currentPeriodEnd': currentPeriodEnd?.toIso8601String(),
      'createdAt': createdAt.toIso8601String(),
      'updatedAt': updatedAt.toIso8601String(),
      'deletedAt': deletedAt?.toIso8601String(),
      'org': org.toJson(),
      'plan': plan.toJson(),
    };
  }

  OrgSubscription copyWith({
    String? id,
    String? orgId,
    String? planId,
    String? status,
    String? stripeCustomerId,
    String? stripeSubscriptionId,
    DateTime? currentPeriodEnd,
    DateTime? createdAt,
    DateTime? updatedAt,
    DateTime? deletedAt,
    Organization? org,
    Plan? plan,
  }) {
    return OrgSubscription(
      id: id ?? this.id,
      orgId: orgId ?? this.orgId,
      planId: planId ?? this.planId,
      status: status ?? this.status,
      stripeCustomerId: stripeCustomerId ?? this.stripeCustomerId,
      stripeSubscriptionId: stripeSubscriptionId ?? this.stripeSubscriptionId,
      currentPeriodEnd: currentPeriodEnd ?? this.currentPeriodEnd,
      createdAt: createdAt ?? this.createdAt,
      updatedAt: updatedAt ?? this.updatedAt,
      deletedAt: deletedAt ?? this.deletedAt,
      org: org ?? this.org,
      plan: plan ?? this.plan,
    );
  }
}

File: server/src/routes/org-subscription.ts
import { Elysia, t } from "elysia";
import { authMiddleware } from "../middleware/auth";
import { orgSubscriptionService } from "../services/orgsubscription";
import { 
  OrgSubscriptionPlainInputCreate, 
  OrgSubscriptionPlainInputUpdate 
} from "../../generated/prismabox/OrgSubscription";
import { regionMiddleware } from "../middleware/region";

export const orgSubscriptionRoutes = new Elysia({ prefix: "/org-subscriptions" })
  .use(authMiddleware)
  .use(regionMiddleware)

  /**
   * GET /org-subscription
   * Retrieves all OrgSubscription with pagination and basic filtering.
   */
  .get("/", async ({ orgId, db, query }) => {
    const { page = "1", limit = "20", ...where } = query as any;
      if (orgId) where.orgId = orgId;
    const result = await orgSubscriptionService.withDB(db as any).getAll({
      where,
      skip: (parseInt(page) - 1) * parseInt(limit),
      take: parseInt(limit),
      orderBy: { createdAt: "desc" },
      include: { plan: true }
    });
    const data = result.data.map((sub: any) => ({
      id: sub.id,
      orgId: sub.orgId,
      plan: sub.plan?.key || sub.planId,
      planId: sub.planId,
      status: sub.status,
      stripeCustomerId: sub.stripeCustomerId,
      stripeSubscriptionId: sub.stripeSubscriptionId,
      currentPeriodEnd: sub.currentPeriodEnd,
      startDate: sub.createdAt,
      amount: sub.plan?.priceMonthlyCents || 0,
      currency: "USD",
      createdAt: sub.createdAt,
      updatedAt: sub.updatedAt,
      limits: sub.plan?.limits || {},
    }));
    return { data, total: result.total };
  }, {
    query: t.Partial(t.Object({
      page: t.Optional(t.String()),
      limit: t.Optional(t.String()),
      orgId: t.Optional(t.String()),
    }))
  })

  /**
   * POST /org-subscription
   * Creates a new OrgSubscription.
   */
  .post("/", async ({ orgId, db, body, set }) => {
    const data = await orgSubscriptionService.withDB(db as any).create(body);
    set.status = 201;
    return { data };
  }, {
    body: OrgSubscriptionPlainInputCreate
  })

  /**
   * GET /org-subscription/:id
   * Retrieves a single OrgSubscription by ID.
   */
  .get("/:id", async ({ orgId, db, params, set }) => {
    const data = await orgSubscriptionService.withDB(db as any).getById(params.id);
    if (!data) {
      set.status = 404;
      return { error: "OrgSubscription not found" };
    }
    return { data };
  }, {
    params: t.Object({ id: t.String() })
  })

  /**
   * PATCH /org-subscription/:id
   * Updates an existing OrgSubscription.
   */
  .patch("/:id", async ({ orgId, db, params, body, set }) => {
    try {
      const data = await orgSubscriptionService.withDB(db as any).update(params.id, body);
      return { data };
    } catch (e) {
      set.status = 404;
      return { error: "OrgSubscription not found or update failed" };
    }
  }, {
    params: t.Object({ id: t.String() }),
    body: OrgSubscriptionPlainInputUpdate
  })

  /**
   * DELETE /org-subscription/:id
   * Deletes a OrgSubscription.
   */
  .delete("/:id", async ({ orgId, db, params, set }) => {
    try {
      await orgSubscriptionService.withDB(db as any).delete(params.id);
      return { success: true, message: "OrgSubscription deleted successfully" };
    } catch (e) {
      set.status = 404;
      return { error: "OrgSubscription not found or already deleted" };
    }
  }, {
    params: t.Object({ id: t.String() })
  });
